joshua/joshua_model.py

```diff
@@ -563,25 +569,91 @@ def set_versionstamped_key(tr, prefix, suffix, value):
                     len(prefix) + 3), value)
 
 
-def _increment(tr, ensemble_id, counter):
+def _increment(tr : fdb.Transaction, ensemble_id : str, counter : str) -> None:
     tr.add(dir_all_ensembles[ensemble_id]['count'][counter], ONE)
 
 
-def _add(tr, ensemble_id, counter, value):
+def _add(tr : fdb.Transaction, ensemble_id : str, counter : str, value : int) -> None:
     byte_val = struct.pack("<Q", value)
     tr.add(dir_all_ensembles[ensemble_id]['count'][counter], byte_val)
 
 
-def _get_snap_counter(tr, ensemble_id, counter):
+def _get_snap_counter(tr : fdb.Transaction, ensemble_id : str, counter : str) -> int:
     value = tr.snapshot.get(dir_all_ensembles[ensemble_id]['count'][counter])
     if value == None:
         return 0
     else:
         return struct.unpack("<Q", b'' + value)[0]
 
 
+def _get_snap_max_runs(tr : fdb.Transaction, ensemble_id : str) -> int:
+    """Read max_runs for this ensemble at snapshot isolation. Returns a default of 0 if max_runs is not set"""
+    result = tr.snapshot.get(dir_all_ensembles[ensemble_id]['properties']['max_runs'])
+    if result == None:
+        return 0
+    value, = fdb.tuple.unpack(result)
+    return value
+
+def _get_snap_timeout(tr : fdb.Transaction, ensemble_id : str) -> int:
+    """Read timeout for this ensemble at snapshot isolation. Returns a default of 5400 if timeout is not set"""
+    result = tr.snapshot.get(dir_all_ensembles[ensemble_id]['properties']['timeout'])
+    if result == None:
+        return 0
```

Comment:
Should this be 5400 (default value)?

Generally, this function is very similar to `_get_snap_max_runs`. So maybe refactor to read all keys within range `dir_all_ensembles[ensemble_id]` and return a dictionary, so that the caller can try get the property, such as `max_runs` or `timeout` with default values.

Reply:
Done